filament/backend/include/private/backend/JobQueue.h

```diff
@@ -0,0 +1,262 @@
+ */
+class JobQueue : public std::enable_shared_from_this<JobQueue> {
+    struct PassKey {};
+public:
+    using Job = utils::Invocable<void()>;
+    using JobId = uint32_t;
+    using Ptr = std::shared_ptr<JobQueue>;
+    static constexpr JobId InvalidJobId = std::numeric_limits<JobId>::max();
+
+    /**
+     * Creates an instance of JobQueue. Users should call this to create one.
+     * @return An instance of JobQueue
+     */
+    static Ptr create() {
+        return std::make_shared<JobQueue>(PassKey{});
+    }
+
+    explicit JobQueue(PassKey) {} // This can be created only via `create()`
+
+    /**
+     * Pushes a new job into storage.
+     *
+     * @param job The function/lambda to be executed.
+     * @param preIssuedJobId The previously issued job ID where this job is assigned to.
+     *  If the value is `InvalidJobId`, a new job ID is generated internally.
+     * @return A new job ID. If a `preIssuedJobId` is provided, that specific ID is returned instead.
+     */
+    JobId push(Job job, JobId preIssuedJobId = InvalidJobId);
+
+    /**
+     * Retrieves the next job from storage.
+     *
+     * @param shouldBlock If true (typically used by ThreadWorker), waits for a job and returns it.
+     *  If false (typically used by AmortizationWorker), tries retrieving a job. But may return an
+     *  empty job if there's no pending job.
+     * @return The next job. If an empty job is returned, it has different meaning depending on the
+     *  `shouldBlock` value:
+     *    - true: Stop processing (shutting down).
+     *    - false: No jobs currently in queue.
+     */
+    Job pop(bool shouldBlock);
+
+    /**
+     * Retrieves a batch of next jobs from storage. Always non-blocking.
+     *
+     * @param maxJobsToPop The maximum number of jobs to retrieve.
+     *  If < 0, retrieves all pending jobs.
+     * @return A FixedCapacityVector<Job> containing the retrieved jobs.
+     * Returns an empty vector if the queue is empty.
+     */
+    utils::FixedCapacityVector<Job> popBatch(int maxJobsToPop);
+
+    /**
+     * Generate a new job ID. This newly generated ID is meant to be used for the `preIssuedJobId`
+     * parameter of `push` method.
+     *
+     * @return A new job ID.
+     */
+    JobId issueJobId() noexcept;
+
+    /**
+     * Cancels a job by its ID.
+     *
+     * @param jobId The job ID to cancel.
+     * @return true if the job was found and cancelled, false otherwise.
+     */
+    bool cancel(JobId jobId) noexcept;
+
+    /**
+     * Signals the storage to shut down, after which no further jobs can be added using the `push`
+     * method.
+     */
+    void stop() noexcept;
+
+private:
+    JobQueue(const JobQueue&) = delete;
+    JobQueue& operator=(const JobQueue&) = delete;
+
+    std::mutex mStorageMutex;
+    std::condition_variable mStorageCondition;
+    std::unordered_map<JobId, Job> mJobsMap;
+    std::queue<JobId> mJobOrder;
+    JobId mNextJobId = 0;
+    bool mIsStopping = false;
```

Comment:
I was confused about what `mIsStopping` means.  Does it mean that you'll no longer issue job ids?   If so, can you add a comment somewhere about that or maybe rename the variable.  Cause I had assumed that you wouldn't be able to pop as well, but apparently you can still pop.